### DISCUSSION TITLE
Miflora Sensors Dups in Home Assistant

### DISCUSSION THREAD

**Discussion by alexbell62** [mKeRix/room-assistant#357] [Q&A]
Hi, 
Is there a way to set up Xiaomi Miflora sensors in room assistant config on multiple raspberry pi zero w the way so its not duplicated in home assistant? 
Currently after configuring miflora sensors on a few raspberry pi zero ws im getting these duplicates in home assistant:
sensor.avocado_humidity
sensor.avocado_humidity_2
sensor.avocado_humidity_3
sensor.avocado_humidity_4

What is the best way to combine them into one sensor and is there a way to set up the room assistant config to do it?

Thanks.
Alex

**Comment by JesusRo** [comment]
Hi Alex,
I found my self in the same scenario than you, but with other xiaomi sensors (humidity/temp LYWSD03MMC)
I ended with a flow on node-red reading all those  `sensor.*_humidity_*` messages and posting them under unified sensor via MQTT too.

I understand that the idea is not to read all sensor from all room-assistant nodes at the same time, but for me it was convenient. Some of the sensor are on places that sometimes report well (because of distance, blocks, wall, interferences...) and others not, to the HA Pi that was the only one before with the classic xiaomi integration.

I leave my flow below but, this is just "my workaround":

```json
[{"id":"773ae754.657078","type":"mqtt out","z":"fedb253d.698188","g":"9209262c.db8a58","name":"","topic":"","qos":"","retain":"","broker":"30ba6ef7.37e75a","x":1130,"y":120,"wires":[]},{"id":"2fc2d861.e83b3","type":"delay","z":"fedb253d.698188","g":"9209262c.db8a58","name":"","pauseType":"rate","timeout":"5","timeoutUnits":"seconds","rate":"1","nbRateUnits":"1","rateUnits":"second","randomFirst":"1","randomLast":"5","randomUnits":"seconds","drop":false,"x":990,"y":120,"wires":[["773ae754.657078"]]},{"id":"42ee31b9.2e311","type":"switch","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Not restarted","property":"payload","propertyType":"msg","rules":[{"t":"neq","v":"offline","vt":"str"}],"checkall":"true","repair":false,"outputs":1,"x":830,"y":120,"wires":[["2fc2d861.e83b3"]]},{"id":"4b1fe5e6.d945d4","type":"function","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Set Xiaomi topic","func":"msg.topic=msg.topic.replace(/^room-assistant\\/sensor\\/[a-z]*-/i, \"room-assistant/sensor/xiaomi-\")\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","x":620,"y":80,"wires":[["42ee31b9.2e311"]]},{"id":"fac7d645.2e08b8","type":"function","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Set Xiaomi stuffs","func":"//Topic\nmsg.topic=msg.topic.replace(/^homeassistant\\/sensor\\/room-assistant\\/[a-z]*-/i, \"homeassistant/sensor/room-assistant/xiaomi-\")\n\n//In Payload\n//Name\nmsg.payload=msg.payload.replace(/\"name\":\"/ig,'\"name\":\"Xiaomi ')\n//Topics for data\nmsg.payload=msg.payload.replace(/room-assistant\\/sensor\\/[a-z]*-/gi,\"room-assistant/sensor/xiaomi-\")\n//id\nmsg.payload=msg.payload.replace(/\"unique_id\":\"room-assistant-[a-z]*-/i,'\"unique_id\":\"room-assistant-xiaomi-')\n//Identifiers\nmsg.payload=msg.payload.replace(/\"identifiers\":\"[a-z0-9]*/i,'\"identifiers\":\"Xiaomi')\n//Model\nmsg.payload=msg.payload.replace(/\"model\":\"[a-z0-9]*/i,'\"model\":\"Xiaomi')\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","x":630,"y":160,"wires":[["42ee31b9.2e311"]]},{"id":"81b3f850.9903f","type":"switch","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Only known","property":"topic","propertyType":"msg","rules":[{"t":"cont","v":"temperature","vt":"str"},{"t":"cont","v":"humidity","vt":"str"},{"t":"cont","v":"battery","vt":"str"}],"checkall":"true","repair":false,"outputs":3,"x":430,"y":80,"wires":[["4b1fe5e6.d945d4"],["4b1fe5e6.d945d4"],["4b1fe5e6.d945d4"]]},{"id":"4c45fb21.547d94","type":"switch","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Only known","property":"topic","propertyType":"msg","rules":[{"t":"cont","v":"temperature","vt":"str"},{"t":"cont","v":"humidity","vt":"str"},{"t":"cont","v":"battery","vt":"str"}],"checkall":"true","repair":false,"outputs":3,"x":430,"y":160,"wires":[["fac7d645.2e08b8"],["fac7d645.2e08b8"],["fac7d645.2e08b8"]]},{"id":"4af1468d.340738","type":"switch","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Not Xiaomi","property":"topic","propertyType":"msg","rules":[{"t":"regex","v":"^((?!xiaomi).)*$","vt":"str","case":true}],"checkall":"true","repair":false,"outputs":1,"x":270,"y":80,"wires":[["81b3f850.9903f"]]},{"id":"329d64f1.e15b64","type":"switch","z":"fedb253d.698188","g":"9209262c.db8a58","name":"Not Xiaomi","property":"topic","propertyType":"msg","rules":[{"t":"regex","v":"^((?!xiaomi).)*$","vt":"str","case":true}],"checkall":"true","repair":false,"outputs":1,"x":270,"y":160,"wires":[["4c45fb21.547d94"]]},{"id":"58797c10.c7bedc","type":"mqtt in","z":"fedb253d.698188","g":"9209262c.db8a58","name":"room-sensors","topic":"room-assistant/sensor/#","qos":"2","datatype":"auto","broker":"30ba6ef7.37e75a","x":110,"y":80,"wires":[["4af1468d.340738"]]},{"id":"8b4974dc.6eca08","type":"mqtt in","z":"fedb253d.698188","g":"9209262c.db8a58","name":"room-entities","topic":"homeassistant/sensor/room-assistant/#","qos":"2","datatype":"auto","broker":"30ba6ef7.37e75a","x":110,"y":160,"wires":[["329d64f1.e15b64"]]},{"id":"30ba6ef7.37e75a","type":"mqtt-broker","name":"HA","broker":"localhost","port":"1883","clientid":"","usetls":false,"compatmode":false,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closeQos":"0","closePayload":"","willTopic":"","willQos":"0","willPayload":""}]
```

Cheers

**Comment by alexbell62** [comment]
Super! Thanks for this!

I think another option is a sensor template that combines all instances in one?
And sounds like atm cant be done in room assistant.

**Comment by alexbell62** [comment]
hi @JesusRo 
It looks like Room assistant integration missing some bits, here are some more details: https://github.com/mKeRix/room-assistant/issues/361.
Hmmm is there a way to make it work?

Also, i wonder if room assistant will work in parallel with one of these:
- https://github.com/andrewjfreyer/monitor
- https://github.com/ChristianKuehnel/plantgateway
- https://github.com/ThomDietrich/miflora-mqtt-daemon

I read somewhere that the room assistant locks Bluetooth of the device when it runs so not sure if these can run simultaneously?

thanks,
Alex

**Comment by JesusRo** [comment]
Hey @alexbell62,
About sensor template stuff, I think it should do the same I do in node-red, it´s just I´m very bad with them xD 

Regarding the metrics of xiaomi sensors, I think each model/version reports differently. In instance, mine ones report temperature and humidity every 10 mins, and battery level every 30 if I remember correctly.
For Mi Flora, I saw some remarks on the [documentation](https://www.room-assistant.io/integrations/xiaomi-mi.html#supported-devices). `requires up-to-date firmware, tested with v3.2.2`

The xiaomi sensor integration I was using before was [xiaomi_bt](https://github.com/custom-components/ble_monitor#supported-sensors) (referenced also on room-assistant docs) and actually I had them both working together (in HASS.io) fine for a while but only BLE, never used active bluethoot scanning. You can give it a try
Also take a look on the devices/models to find yours, maybe it´s not publishing all the info

**Comment by alexbell62** [comment]
Huge thanks for this!
 i connect it to Xiaomi app according to what i found online this will upgrade firmware automatically.

Thanks for the custom component link. I also used in-built HA component before, this one: https://www.home-assistant.io/integrations/miflora/
...but it stopped sending the data a day or so ago. Now i have all these raspberry pi zero w around my house thanks to room assistant but main HA raspberry pii is in the loft away from everything. So i was hoping to increase reliability and make every raspberry pi zero w convert mi flora ble signals into mqtt so they are never lost because of the distance.

Hmm, just not sure how to do it and where to start.

**Comment by mKeRix** [comment]
It's currently not implemented in room-assistant since I wasn't aware that this is a use case - so far I just assumed people would configure the Xiaomi Mi stuff on one instance only, whichever is the closest to the device. I wouldn't want to change the default behavior here, but I could imagine opening this up as a config option. Feel free to open a feature request for this!

**Comment by PeteBa** [comment]
I jumped the gun  8-) and just pushed a PR (#362) that swaps  `distributed` to TRUE so that any one xiaomi device can be picked up by any node and be considered as the same entity. To be honest, I thought this was probably what was meant originally and have probably missed the use-case for `distributed` as false. I can imaging it would be nice to associate a plant with a "room" but I think many people may, like me, move their mifloras around the house to different plants all the time. It also means that if one node goes down then the sensor still gets updated via another node. Obviously, happy to adjust the PR as needed.

**Reply by mKeRix** [reply]
Yeah, I also gave it another thought and realized that the use case for having it as not distributed is rather slim, while having it as distributed allows you to move the sensors/plants anywhere around the house. Thanks for improving on this :)

**Reply by alexbell62** [reply]
wow! That's awesome news thanks so much @PeteBa and @mKeRix for fixing and working on this!
Is this change planned to be merged/released in a new version any time soon? Just wondering if I should just wait for it a bit or upgrade from @PeteBa 's branch?

**Reply by mKeRix** [reply]
I think I'll put this out as part of the current beta releases over the course of the coming week. Usually I don't like leaving these things unreleased for too long, but here I need to make some small adjustments in the core for it to work perfectly and just didn't have the time/energy in the last few days.

**Reply by mKeRix** [reply]
The new beta version containing these changes is now out. :)

**Reply by PeteBa** [reply]
Fantastic, many thanks for including this!

Any chance you can describe the `stateLocked` property and when it should be used? I'm trying to understand the use-case for when both `stateLocked` and `distributed` would both be set to true. Again, thanks for this.

**Reply by mKeRix** [reply]
I probably should've put a JSDoc on the constructor as well 😅 `stateLocked` determines whether a distributed entity can receive state updates from any node or just the leader node. Both being true was essentially the previous behavior for distributed entities. This is used heavily in the BLE and BT Classic integrations: these integrations interchange messages between instances containing RSSI data. Each instance then keeps its own cache of the data it saw and calculates a state from it (= which room the device is in). In an ideal world all these instances would have the exact same data locally and therefore also calculate the same state - but in the real world this isn't always the case (new instances connecting, messages getting lost...). To prevent weird state switches in HA due to this only the leader node may update the entity (`stateLocked` = true).

**Reply by alexbell62** [reply]
Yay great news! Thanks for the merge! 
Going to test it next few days 👍   

Btw is there a way to use Ansible to install different versions of nodejs on different hosts?  

And also can i use both BLE and Bluetooth classic on raspberry pi zero w simultaneously: one for Xiaomi plant sensor another for to track watches?

**Reply by alexbell62** [reply]
I've been testing it today and it looks cool so far! Just one entry per plant/measure instead of many! Awesome, thanks again!

**Reply by mKeRix** [reply]
With Ansible in general yes, but not in the playbook as I have put it up here on GitHub. Shouldn't require too many modifications though, if you want to look into it. :)

And yes, with the most recent versions of room-assistant you can run both BT Classic and any BLE integration simultaneously without any extra hardware.

**Reply by alexbell62** [reply]
haha, i will try and see how far it goes, thanks! 


>can run both BT Classic and any BLE integration simultaneously

Feels like Christmas arrived earlier lol
Such great news, going to test this week!
